### DISCUSSION TITLE
Add periodic account renewal

### DISCUSSION THREAD

**Discussion by TiCampbell** [Get-Your/Get-Your#204] [Ideas]
Per some program requirements, we'll need a periodic 'renewal' period, where active users will be required to update their information (using the mechanism of Get-Your/Get-Your#69, but this account renewal won't be voluntary).

We'll need both an app-side and business-side process for this, including how to keep the user's data but flag that it needs to be changed or confirmed before the 'renewal' is complete (an `up_to_date` Boolean flag?).

Once information is updated, the renewal should be marked 'closed' and the period reset. All programs for which they qualify will continue or change as needed (with email notifications as addressed in Get-Your/Get-Your#37) with nothing needed from the program coordinators (which means they no longer need to be concerned with anything renewal-related).

**Comment by TiCampbell** [comment]
Recreation has a separately-triggered annual renewal process (based on when a pass is purchased, not when they're initially enrolled), but maybe a global renewal process (and hopefully less frequently than annual) would suffice and we could drop that requirement.

**Comment by TiCampbell** [comment]
This will initially be an annual process (to match what's been done at the City previously) and we're going to match the timeframe to when AMI values are updated with the current year (each May-June). This is convenient for notifications because the renewal will include updating income ranges and therefore we won't need another email notification for when the ranges are updated.

This combination of 'renewal' and 'AMI range updates' needs to be noted somewhere (at least in the code), so that if (when) the renewal period is increased, we'll know to add annual email reminders for users to update their income information to see if they qualify for additional programs.

**Reply by TiCampbell** [reply]
Since AMI is no longer included in the app, this requirement is moot. A January renewal would make it easiest to mark a user as "<this year's> renewal", but the timeframe matters less once we have a `last_renewed_at` timestamp.

**Comment by TiCampbell** [comment]
Note that I added a stub for `renewal_mode` as a Boolean input to `app.views.application.take_usps_address()` for completion - this needs to be expanded in the same way as `update_mode`.

See PR #184 and commit 5a7eb70f, if still relevant.

**Comment by TiCampbell** [comment]
EDIT: THE THINKING ON THIS HAS CHANGED - see later comments.

@penoptech I figured this is larger than a single issue, so I moved it to a discussion. We've been asked to move this to top priority for new functionality (above #164) because of the 2023 Grocery Tax Rebate renewal.

I'm using 'affected user' below because it might be nice to be able to target only a subset of users. Also, this year's GTR renewal will only be for GTR participants who enrolled last year, so I'll need to have a targeted approach for that anyway.

Here's what I'm picturing (with our current setup):
1. I'll kick off the renewal process (either manually or scheduled), which would
    1. Un-enroll each affected user from the affected programs (TBD another time - GTR will certainly be included since it's an annual benefit, but the other programs hopefully can be less frequently)
    2. Reset a new `is_renewed` flag (similar to `is_updated`, but this would default to `True` for new users)
2. (We would send out an email notification (included here for reference only - not covered within the app))
3. When an affected user logs in, if `has_viewed_dashboard==True`, take them to a modified dashboard with just a message that they need to renew their account to continue getting benefits and a link to start the renewal process. If `has_viewed_dashboard==False` we can assume they haven't logged in before and this step can be skipped
4. The link to start the renewal process (or `has_viewed_dashboard==False`) will bring them back to the beginning of the application (with `renewal_mode` set in the session variables). The application can have the same functionality as `update_mode`, which pre-save signals record any changes in the historical tables

    > It would be great to have each page include the ability for the 'continue' button to default to display 'confirm', but change to 'update' if anything is changed in the form
6. Once they reach the dashboard, `is_renewed` is set, at which point the process kicks off as if it were a new applicant (with some notes to the program coordinators to specify that this is a '<this year> renewal'. It would also be helpful to have a `last_renewed_at` timestamp in the `User` model that would be updated as `is_renewed` is set.

    > The `broadcast` page will need to either send a different email/text or have communications disable for renewals

Let me know what you think and also any details I'm missing.

**Reply by penoptech** [reply]
@TiCampbell, I was thinking about how we can tell business users that a person they're seeing in an extract is someone renewing their enrollment vs a brand new enrollment. At first I thought we could just compare the `last_renewed_at` of the user with the `applied_at` for the IQ program, but the problem with this approach is that someone might be enrolling in an IQ program for the first time after a renewal, so that comparison won't necessarily always be correct.

Does it make sense to apply your `is_renewed` and `last_renewed_at` to the IQ programs as well? Except maybe we call it `has_reapplied` and `last_applied_at`. If we have that data, I think it'd be possible to compare `last_renewed_at` with `last_applied_at` to see if someone is a brand new enrollment vs reenrolling in an IQ program.

What do you think?

**Reply by TiCampbell** [reply]
@penoptech That's a great call, but like we discussed yesterday, we may run into issues after multiple renewal periods. I like the `has_renewed` bool with an `IQProgramHist` model that will preserve original timestamps. That way we can use the existing `applied_at` and `enrolled_at` timestamps for each renewal without losing the original (or previous-renewal) data.

**Comment by TiCampbell** [comment]
Per our conversation with Nina today, it's confirmed we're going to go with a 'rolling' renewal (as opposed to the 'batch' renewal I described above), where each user will have the renewal process kicked off automatically one year after they completed their initial application (or previous renewal). We still need to see if this can be done within Django (even for users that haven't logged in in a long time) or if it should be a separate (nightly?) process.

We also learned that programs have different renewal frequency, so for each affected user we'll want to "un-enroll" only the programs for which they have been enrolled (or have applied for?) for longer than the renewal frequency (e.g. for an annual renewal process, generic Program B with a two-year renewal frequency will *not* be un-enrolled if this is the first year since initial enrollment).

> Along the lines of differing renewal frequencies, we'll probably want to have the overall renewal frequency (which we've been assuming was 'annual') be the minimum of all programs' renewal frequencies.

One of this issues this brings up is how do users view their dashboard to see that they're still renewed in Program B if at renewal time they're no longer enrolled in any Eligibility Program? I think we need a way for users to reach the dashboard without selecting any programs - maybe a 'skip for now' button that pops up a modal warning that they're not eligible for any IQ Programs until they come back and select an Eligibility Program.

**Reply by TiCampbell** [reply]
The 'skip for now' is actually now a viable option, since users can now update their information, which could presumably be used to select an Eligibility Program (although this isn't currently an option in Settings), and a bitmap (such as referenced in #53) can be used for application/renewal progress (or a different bitmap for each) to remove the issue of `what_page()` trying to take the user back to Eligibility Program selection.

**Reply by penoptech** [reply]
@TiCampbell, here's are a couple potential options for us to run periodic tasks directly in Django. My only issue with them is that they haven't really been updated recently. I wonder if Postgres has an option for building and running scheduled tasks, like the SSIS packages on SQL Server

https://django-cron.readthedocs.io/en/latest/installation.html
https://django-q.readthedocs.io/en/latest/

There's also a v2 of django-q, but I need to look through the documentation.

Whatever we do I just want to make sure it's somewhat light weight and portable. I've used Celery and RabbitMQ for periodic tasks in the past and it's WAY to heavy for what we're trying to accomplish.

Ok. So dug into the `django-q2`, this is actually looking pretty promising. The repo was forked from `django-q`, is actually being maintained, and is compatible with Django 4. I'll start creating a proof of concept with the package to see what it can do.

https://django-q2.readthedocs.io/en/master/index.html

**Comment by penoptech** [comment]
@TiCampbell, do we need to tell the Grocery Tax Rebate folks when someone is no longer eligible based on an update to their address or proof of AMI, or do they only issue checks based on the extract they receive from Get FoCo?

And maybe using GTR as the example is too specific... It might be better to think about this more generically. Should we be giving program coordinators some kind of change report (additions and subtractions for enrollments to their program). There's quite a bit to think through if this is necessary, so maybe the focus is just on GTR since they're the only program that has a renewal if I was understanding Nina correctly.

**Reply by TiCampbell** [reply]
@penoptech Yeah, we've been talking kind of specifically about GTR because they'll be affected immediately, since they're the only program with 1-year renewal interval (the others are longer). A generic solution would definitely be better though - especially because GTR is an intentional checks-mailed-out process (as opposed to the Recreation 'a user is in the system whenever they come in to purchase a pass').

I think it makes the most sense for us (for a generic solution) to specify every 'state change' of a user, to keep our workflow independent of however the coordinators want to do it. So we'd have these user types:

- New
- Updated information
- Renewal
- ~~Removal (when a user no longer meets the eligibility criteria)~~

Each coordinator could then ignore whatever they don't need (e.g. GTR wouldn't care about 'removal' since they're gathering applicable users each year).

**Reply by TiCampbell** [reply]
I removed the 'removal' option from the user types; I think this purports knowing how a particular backend system works, which we don't want to do.

**Comment by TiCampbell** [comment]
Now that the 2023 GTR renewal has been released and is underway, this discussion is effectively complete. I moved any relevant bits over to a discussion of a more-robust process for 2024 and onward renewals in #228.
